src/rhsmlib/cloud/README.md

```diff
@@ -1,8 +1,189 @@
+            return True
+        # In other cases return False
+        return False
+
+    def is_likely_running_on_cloud(self):
+        """
+        Return non-zero value, when the machine is virtual machine and it is running on Foo
+        hypervisor and some Foo string can be found in output of dmidecode
+        :return: Float value representing probability that vm is running on Foo
+        """
+        probability = 0.0
+
+        # TODO: Check if this is true for Foo Cloud Provider
+        if self.is_vm() is False:
+            return 0.0
+
+        if 'dmi.chassis.asset_tag' in self.hw_info and \
+                self.hw_info['dmi.chassis.asset_tag'] == 'FOO':
+            probability += 0.3
+
+        # Try to find "Foo" keyword in output of dmidecode
+        found_foo = False
+        for hw_item in self.hw_info.values():
+            if type(hw_item) != str:
+                continue
+            if 'foo' in hw_item.lower():
+                found_foo = True
+        if found_foo is True:
+            probability += 0.3
+
+        return probability
+
+
+    def _get_metadata_from_cache(self) -> Union[str, None]:
+        """
+        TODO: implement something or return None
+        """
+        raise NotImplementedError
+
+    def _get_data_from_server(self, data_type, url):
+        """
+        This method tries to get data from server using GET method
+        :param data_type: string representation of data type used in log messages (e.g. "metadata", "signature")
+        :param url: URL of GET request
+        :return: String of body, when request was successful; otherwise return None
+        """
+        return super(FooCloudProvider, self)._get_data_from_server(data_type, url)
+
+    def _get_metadata_from_server(self) -> Union[str, None]:
+        """
+        Try to get metadata from server
+        :return: String with metadata or None
+        """
+        return super(FooCloudProvider, self)._get_metadata_from_server()
+
+    def _get_signature_from_cache_file(self) -> Union[str, None]:
+        """
+        TODO: implement something or return None
+        """
+        raise NotImplementedError
+
+    def _get_signature_from_server(self) -> Union[str, None]:
+        """
+        Method for gathering signature of metadata from server
+        :return: String containing signature or None
+        """
+        return super(FooCloudProvider, self)._get_signature_from_server()
+
+    def get_signature(self) -> Union[str, None]:
+        """
+        Public method for getting signature (cache file or server)
+        :return: String containing signature or None
+        """
+        return super(FooCloudProvider, self).get_signature()
+
+    def get_metadata(self) -> Union[str, None]:
+        """
+        Public method for getting metadata (cache file or server)
+        :return: String containing metadata or None
+        """
+        return super(FooCloudProvider, self).get_metadata()
```

Comment:
these seems not needed to reimplement in subclasses

Reply:
Again, this is example describing complete API.